### ISSUE TITLE
Service should run as simple, not forking

### ISSUE THREAD

**Issue by amorsen** [dun/conman#36]
First opened as https://bugzilla.redhat.com/show_bug.cgi?id=2174838 but it was recommended to go upstream.

conman currently does not use the systemd journal or the standard systemd service invocation, despite having the necessary code (option F).

This can be fixed by changing the service section of conman.service.in from

[Service]
Type=forking
ExecStart=@sbindir@/conmand -P @runstatedir@/conmand.pid
PIDFile=@runstatedir@/conmand.pid
LimitNOFILE=infinity
Restart=on-abort

to

[Service]
Type=simple
ExecStart=@sbindir@/conmand -F
LimitNOFILE=infinity
Restart=on-abort

conman is in production here with the change to Type=simple, and it works as it should. It works with User=conman as well, if the permissions to /var/log/conman and /var/log/conman.old are changed appropriately. This is a more secure configuration.

**Comment by amorsen** [comment]
Patch at https://gist.github.com/amorsen/b4f01c828d9070a81f690776814ce9e3

**Comment by dun** [comment]
*[My response posted to the above [Bugzilla](https://bugzilla.redhat.com/show_bug.cgi?id=2174838) issue]*

This issue seems to conflate two independent items: running `conmand` as a systemd `type=simple` service, and running `conmand` as a non-privileged user.

`conmand` supports running in the foreground (via the `-F` cmdline option), but that was originally intended for use in development with the assumption that the production daemon would run as a background process. Some of those assumptions haven't been fixed yet; for example, `chdir()` is only called in the daemonization code which performs the double-`fork()`. Also, there is no means to filter the priority of log messages written in the foreground (see `<systemd/sd-daemon.h>`). So I don't plan to switch the service file to `type=simple` until I've revisited startup and logging to better integrate with systemd. Note that a `type=forking` service can still run as a non-privileged user by writing its PID file to a runstatedir created by the systemd `RuntimeDirectory` directive. However...

`conmand` currently runs as the root user, and has done so historically. It can successfully run as a non-privileged user in some configurations. But I'm hesitant making that change upstream without further consideration. For example, access to local serial devices requires elevated privileges, although that could likely be remedied by having `conmand` run as a member of the "dialout" group. I'd need to do more testing, and check if there are other areas where root privileges are assumed or required. Also, the root privileges are mitigated to an extent by Fedora's SELinux policy which constrains `conmand`.

In summary, `type=forking` is better supported in `conmand` at the moment, although plans are to switch to `type=simple` (and a more sandboxed configuration) in the future when time allows.

**Comment by dun** [comment]
`conmand` can write its log messages to `syslog()` via the *conman.conf* `syslog` directive:
```
server syslog="<facility>"
```

Those messages will appear in the systemd journal along with their associated metadata including the syslog *priority* level.

Are you unable to get these messages in the systemd journal using this directive?

**Comment by amorsen** [comment]
Fair.

Switching to RuntimeDirectory and syslog logging is more effort to maintain locally. In this particular use case, the consoles are reached through ssh and having ssh client run as root is not desirable.

I will just keep my service override file, since it works better for my use case.

**Comment by dun** [comment]
I don't see how syslog logging is that much more effort. You need to add a single line to *conman.conf*. The *conman.service* file to run a non-privileged (user=conman) forking daemon would be:
```
[Unit]
Description=ConMan Console Manager
Documentation=man:conmand(8)
After=network.target
After=time-sync.target

[Service]
Type=forking
ExecStart=/usr/sbin/conmand -P /run/conman/conmand.pid
PIDFile=/run/conman/conmand.pid
RuntimeDirectory=conman
LimitNOFILE=infinity
Restart=on-abort
User=conman

[Install]
WantedBy=multi-user.target
```

With `RuntimeDirectory`, systemd manages the creation/destruction of the `/run/conman` directory.

At any rate, I'm glad you found a solution that works for you.
